packages/matchbox/src/components/Slider/Slider.js

```diff
@@ -0,0 +1,177 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classnames from 'classnames';
+import { getRectFor, lerp } from '../../helpers/geometry';
+import { noop, isNotTouchEvent } from '../../helpers/event';
+import { onKey, onKeys } from '../../helpers/keyEvents';
+import { roundToPlaces, trim } from '../../helpers/math';
+import styles from './Slider.module.scss';
+
+function Slider(props) {
+  const { defaultValue, disabled, max, min, onChange, precision, value } = props;
+
+  const [sliderValue, setSliderValue] = React.useState(value || defaultValue || 0);
+  const [pixelOffset, setPixelOffset] = React.useState(0);
+  const [moving, setMoving] = React.useState();
+
+  const sliderRef = React.useRef();
+
+  // Sets internal value when this is a controlled component
+  React.useLayoutEffect(() => {
+    if (value) {
+      setSliderValue(value);
+    }
+  }, [value]);
+
+  // Calculates pixel offset for handle and track when internal value changes
+  React.useLayoutEffect(() => {
+    const rect = getRectFor(sliderRef.current);
+    const trimmedValue = trim(sliderValue, min, max);
+    const absolutePercentage = (trimmedValue + Math.abs(min)) / Math.abs(min - max);
+    setPixelOffset(lerp(0, rect.width, absolutePercentage));
+    if (!isNaN(parseFloat(sliderValue)) && isFinite(sliderValue) && onChange) {
```

Comment:
Hmm not sure what you mean here:
```js
const [sliderValue, setSliderValue] = React.useState(
  !isNaN(parseFloat(value)) && isFinite(value)
    ? value
    : defaultValue
);
```

maybe in the effect?
```js
  React.useEffect(() => {
    if (!isNaN(parseFloat(value)) && isFinite(value)) {
      setValue(value);
    }
  }, [value]);
```

Reply:
Added the check to the effect, which syncs the internal value from the outside

Reply:
Yeah, I meant we should sanitise the value coming from outside only when require instead of parsing, NaN-checking and finite-checking it each time we want to use it.

The effect triggered by `value` looks good 👍